src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -910,6 +914,89 @@ private void CreateFirstSegment(string segmentText)
             this.CreateDynamicPathSegment(null, identifier, parenthesisExpression);
         }
 
+        /// <summary>Binds a <see cref="ODataPathSegment"/> for an escape function segment.</summary>
+        /// <param name="segmentText">The text of the segment.</param>
+        private bool BindSegmentForEscapeFunction(string segmentText)
+        {
+            string identifier;
+            string parenthesisExpression;
+            ExtractSegmentIdentifierAndParenthesisExpression(segmentText, out identifier, out parenthesisExpression);
+
+            if (this.parsedSegments.Count == 0)
+            {
+                if (this.TryCreateSegmentForNavigationSource(identifier, parenthesisExpression))
+                {
+                    return true;
+                }
+
+                if (this.TryCreateSegmentForOperationImport(identifier, parenthesisExpression))
+                {
+                    return true;
+                }

```

Comment:
here looks duplicated a lot of codes, should be just call "CreateFirstSegment"... and in else to call "CreateNexteSegment"?

Reply:
That is what the previous iteration was doing. It added a lot of checks to existing cases. I did not want to degrade existing performance in the hot path.

---
In reply to: [369804646](https://github.com/OData/odata.net/pull/1621#discussion_r369804646) [](ancestors = 369804646)